packages/back-end/test/api/environments.test.ts

```diff
@@ -0,0 +1,775 @@
+}));
+
+jest.mock("../../src/services/auth", () => ({
+  ...jest.requireActual("../../src/services/auth"),
+  getAuthConnection: () => ({
+    middleware: jest.fn(),
+  }),
+}));
+
+jest.mock("../../src/middleware/authenticateApiRequestMiddleware", () => ({
+  ...jest.requireActual(
+    "../../src/middleware/authenticateApiRequestMiddleware"
+  ),
+  __esModule: true,
+  default: jest.fn(),
+}));
+
+let mongod;
+
+describe("environements API", () => {
+  const OLD_ENV = process.env;
+
+  beforeAll(async () => {
+    jest.resetModules();
+    mongod = await MongoMemoryServer.create();
+    const uri = mongod.getUri();
+    process.env.MONGO_URL = uri;
+    getAuthConnection().middleware.mockImplementation((req, res, next) => {
+      next();
+    });
+  });
+
+  afterAll(async () => {
+    await mongoose.connection.dropDatabase();
+    await mongoose.connection.close();
+    await mongod.stop();
+    process.env = OLD_ENV;
+  });
+
+  afterEach(async () => {
+    jest.clearAllMocks();
+    const collections = mongoose.connection.collections;
+    for (const key in collections) {
+      const collection = collections[key];
+      await collection.deleteMany();
+    }
+  });
+
+  it("can list all environments", async () => {
+    authenticateApiRequestMiddleware.mockImplementation((req, res, next) => {
+      req.context = {
+        org: {
+          settings: {
+            environments: [
+              {
+                id: "env1",
+                description: "env1",
+                toggleOnList: true,
+                defaultState: true,
+                projects: ["bla"],
+              },
+              {
+                id: "env2",
+              },
+            ],
+          },
+        },
+        permissions: {
+          canReadMultiProjectResource: () => true,
+        },
+      };
+
+      next();
+    });
+
+    const response = await request(app)
+      .get("/api/v1/environments")
+      .set("Authorization", "Bearer foo");
+
+    expect(response.status).toBe(200);
+    expect(response.body).toEqual({
+      environments: [
+        {
+          id: "env1",
+          description: "env1",
+          toggleOnList: true,
+          defaultState: true,
+          projects: ["bla"],
+        },
+        {
+          id: "env2",
+          description: "",
+          toggleOnList: false,
+          defaultState: false,
+          projects: [],
+        },
+      ],
+    });
+  });
+
+  it("can filter environments", async () => {
+    authenticateApiRequestMiddleware.mockImplementation((req, res, next) => {
+      req.context = {
+        org: {
+          settings: {
+            environments: [
+              {
+                id: "env1",
+                description: "env1",
+                toggleOnList: true,
+                defaultState: true,
+                projects: ["bla"],
+              },
+              {
+                id: "env2",
+              },
+            ],
+          },
+        },
```

Comment:
Lots of copy/pasting in this test files makes it a little hard to follow and maintain.  For example, if we add a new method to req.context that is needed under-the-hood, we will have to update it 13 times in this file.